### PR TITLE
spu/mfc: events

### PR DESCRIPTION
**Pull request by elad335**
Implement events and attempt to fix anything related to it

### CODE REVIEW

**Comment by isJuhn** [comment]
this pr fixes the Trap! in Ratchet and Clank: Into the Nexus [BCES01908] and it now shows an intro but dies before the menu with 
`E {rsx::thread} HLE: Getting the real address failed. (addr=0x80c1a600)
F {rsx::thread} class std::runtime_error thrown: GetAddress(offset=0x80c1a600, location=0x1): RSXIO memory not mapped
(in file Emu\RSX\RSXThread.cpp:55)`
rpcs3_2017-12-17_19-30-16
rpcs3_2017-12-17_19-32-57

However, if I (thanks elad for building) mask the RSX address with 0xFFFFFFF like I detailed in #3387 I am able to get to the menu and start a new game. I even displays the intro cutscene but dies when it should go ingame, the audio is working even though it died.
rpcs3hack_2017-12-17_20-01-16
rpcs3hack_2017-12-17_20-02-15
rpcs3hack_2017-12-17_20-05-08

**Comment by elad335** [comment]
image
hitman sniper gets further

**Comment by Xcedf** [comment]
Ratchet and Clank ToD Demo goes into menu with your rsx branch fix

**Comment by isJuhn** [comment]
but that branch isn't this pr, also I already got it to menus a while ago, but it's a hack so it can't be merged.

**Comment by AniLeo** [comment]
Don't post misleading results.

**Comment by Xcedf** [comment]
Sorrry, removed my comment

**Comment by Zangetsu38** [comment]
Datura regression, passed ingame to nothing
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1566168/RPCS3.log.gz)

**Comment by isJuhn** [comment]
This PR makes Ratchet and Clank a crack in time go past the menus, but it crashes once it is supposed to go ingame
rpcs3_2017-12-18_00-07-15
rpcs3_2017-12-18_00-11-28
rpcs3_2017-12-18_00-14-00

Log: [RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1566182/RPCS3.zip)

**Comment by slashiee** [comment]
Sonic Unleashed [BLES00425] no longer gets to title under SPU Interpreter Precise using this PR.

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1566229/RPCS3.log.gz)

**Comment by JackONeill12** [comment]
Assassins Creed Lady Liberty still regressing.
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/1566250/RPCS3.log)

God of War III also still black screen.
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/1566255/RPCS3.log)

General Performance of Sonic Unleashed is much worse than before.
Interestingly the Frame counter still shows around 30 FPS on this PR and on master but on this PR it's more like 1-2 FPS.

**Comment by GeniusMage** [comment]
**Tales of Vesperia**
- Performance reduced by up to 30% (depends on previous performance: if it was high, it is severly reduced, if it was already low, it doesn't change by much)
- Random MEM Access Violation on MFC Thread (see log)
- Random Segfault (couldn't screenshot that one, will try to do so)
- Random unspecified crash ("RPCS3 has stopped responding")

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1567267/RPCS3.log.gz)

**Comment by Nekotekina** [comment]
Creating a thread isn't going to work, overhead is too big.

**Comment by elad335** [comment]
dunnon ,the other option is to have the check for event in mfc thread, and thats slow things down.
also its asleep most of the time and terminates when the event is signalled or the decrementer stops

File: rpcs3/Emu/Cell/SPUThread.cpp [hunk 1]
```diff
@@ -1520,37 +1514,48 @@ bool SPUThread::set_ch_value(u32 ch, u32 value)
...
 		}
+		else {
+			bool add_thread = (dec_state & dec_upd) == 0;
+			dec_state = (dec_upd + dec_run) | (value >> 31);
 
-		// detect masking unimplemented events
-		if (value & ~SPU_EVENT_IMPLEMENTED)
-		{
-			break;
+			if (add_thread) std::thread (&SPUThread::decrementer_thread ,this).detach();
```
Comment (jjsat) [inline rpcs3/Emu/Cell/SPUThread.cpp]: I suspect this PR adds a lot of slowdown to games because it can create a lot of very short lived threads. At least this is true for Dante's Inferno where the fps tanks down to ~ 1fps. Since every SPE only has one counter, a thread for each one could be created once and reused. Or maybe just one thread to serve the counter events for all SPEs.

Comment (elad335) [reply rpcs3/Emu/Cell/SPUThread.cpp]: seems hard but might worth the effort

**Comment by elad335** [comment]
unknown
Loadable -> Almost Ingame on interpreter
Loadable -> first seconds Ingame than freeze on asmjit

**Comment by Xcedf** [comment]
The recent version of this PR unfixes Trap for Ratchet and Clank Nexus and others which have it on older builds i was able to ran the game from second-third try (Trap happend but one- two times only) now it always happens

File: rpcs3/Emu/Cell/SPUASMJITRecompiler.cpp [hunk 2]
```diff
@@ -309,18 +309,16 @@ inline asmjit::X86Mem spu_recompiler::XmmConst(__m128i data)
 void spu_recompiler::CheckInterruptStatus(spu_opcode_t op)
 {
 	if (op.d)
-		c->lock().btr(SPU_OFF_8(interrupts_enabled), 0);
+		c->lock().btr(SPU_OFF_32(ch_event_stat), 31);
 	else if (op.e)
-	{
-		c->lock().bts(SPU_OFF_8(interrupts_enabled), 0);
+		c->lock().bts(SPU_OFF_32(ch_event_stat), 31);
```
Comment (kd-11) [inline rpcs3/Emu/Cell/SPUASMJITRecompiler.cpp]: Indentation for this whole function is just wrong

**Comment by Xcedf** [comment]
Made some tests for older version of this PR
https://github.com/RPCS3/rpcs3/commit/5d83412af8f933bc2bf05d9295f3215014f53fbe this one the latest which able to run R&C games that has Trap on master
the next one https://github.com/RPCS3/rpcs3/commit/d72c9f187659ef68197a9b13e74818b4fcf8d058
already unable to run them. Found a few difference betwen them but the one responsible for R&C games is in SPU_RdEventStat function, i'm not a coder so i can't pass it in your new version, please take a look

**Comment by elad335** [comment]
Rest in peace. lol

I might use the bisled commit later on though.
